### PR TITLE
Fenwick Tree

### LINKED ISSUE #6671: Fenwick Tree
### Description

Documentation for Fenwick Tree Introduction and Implementation with Point and Range Updates

### Domain

C/CPP

### Type of Contribution

Documentation

### Code of Conduct

- [X] I follow [Contributing Guidelines](https://github.com/girlscript/winter-of-contributing/blob/main/.github/CONTRIBUTING.md) & [Code of conduct](https://github.com/girlscript/winter-of-contributing/blob/main/.github/CODE_OF_CONDUCT.md) of this project.

### PR DESCRIPTION
**Pull request by shreya024**
<hr>

## Description 📜

Added code and Documentation for Fenwick Tree Introduction and Implementation with Point and Range Updates. 
Fixes #6671 

<hr>

## Type of change 📝

- [ ] Au#dio (Should be in mp3 format Includes speech clarity, Concise ,Low distortion)
- [ ] Vi#deo (Animations, screen-recordings, presentations and regular explanatory films are all possibilties etc)
- [x] Documentation (Content Creation in the form of codes or tutorials)
- [ ] Other (If you choose other, Please mention changes below) 

<hr>

## Domain of Contribution 📊

- [ ] Android Dev #(Flutter)
- [ ] Android Dev #(Java)
- [ ] Android Dev #(Kotlin)
- [ ] Backend Dev #(Java)
- [ ] Backend Dev #(.NET)
- [ ] Backend Dev #(PHP)
- [ ] Backend Dev #(Python)
- [x] C/CPP
- [ ] Competitive #Programming
- [ ] Cyber #Security
- [ ] DS#A
- [ ] Data#base
- [ ] Datascience with #Python
- [ ] Datascience with #R
- [ ] Frontend Dev #HTML/CSS/JS
- [ ] Frontend Dev #React/Angular/Vue
- [ ] Go#lang
- [ ] Interview #Prep
- [ ] Java_#Domain
- [ ] JavaScript_#Domain
- [ ] ME#RN
- [ ] Machine #Learning
- [ ] Open #Source
- [ ] Python_#Domain
- [ ] Ru#st
- [ ] Statis#tics
- [ ] UI/#UX

<hr>

## Checklist ✅

- [x] I follow [Contributing Guidelines](https://github.com/girlscript/winter-of-contributing/blob/main/.github/CONTRIBUTING.md) & [Code of conduct](https://github.com/girlscript/winter-of-contributing/blob/main/.github/CODE_OF_CONDUCT.md) of this project.
- [x] I have performed a self-review of my own code or work.
- [x] I have commented my code, particularly in hard-to-understand areas.
- [x] My changes generates no new warnings.
- [x] I'm GWOC'21 contributor

<hr>

## Screenshots / Gif (Optional) 📸

<hr>

### CODE REVIEW

**Comment by shreya024** [comment]
@Suvraneel Kindly review my PR

File: C_CPP/Data Structures/User Defined Data Structures/Fenwick Tree/Fenwick_Tree.cpp [hunk 1]
```diff
@@ -0,0 +1,88 @@
...
+                cout<<arr[i]<<"\n";
+            }        
+            break;
+       
+        default:
+            break;
+        }
+    }
+    return 0;
+}
```
Comment (Suvraneel) [inline C_CPP/Data Structures/User Defined Data Structures/Fenwick Tree/Fenwick_Tree.cpp:88]: Add new line at EOF

File: C_CPP/Data Structures/User Defined Data Structures/Fenwick Tree/Fenwick_Tree.md [hunk 2]
```diff
@@ -0,0 +1,52 @@
+<h1 align="center">  Fenwick Tree or Binary Indexed Tree </h1>
+
+A Fenwick Tree is a user-defined data structure that helps to calculate Prefix sums or cumulative frequency tables in a very efficient manner.It is a representational form of an array and can query or update elements in O(logN) time complexity.It is also known as a Binary Indexed Tree as it makes use of bit manipulation concepts for implementation purposes.
+
+Need:
```
Comment (Suvraneel) [inline C_CPP/Data Structures/User Defined Data Structures/Fenwick Tree/Fenwick_Tree.md:5]:
```suggestion
Need:
```

Format these subheadings (Use ## to sub-head or ** to embolden)

File: C_CPP/Data Structures/User Defined Data Structures/Fenwick Tree/Fenwick_Tree.md [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+
+- We can perform both Query and Update operations even in a large array in O(logN) time using Fenwick Trees.
+- Fenwick Trees require less space as compared to Segment Tress.
+- They have an easy implementation code which is useful during various programming contests.
+
+Terminology:
```
Comment (Suvraneel) [inline C_CPP/Data Structures/User Defined Data Structures/Fenwick Tree/Fenwick_Tree.md:11]: "

File: C_CPP/Data Structures/User Defined Data Structures/Fenwick Tree/Fenwick_Tree.md [hunk 4]
```diff
@@ -0,0 +1,52 @@
+<h1 align="center">  Fenwick Tree or Binary Indexed Tree </h1>
+
+A Fenwick Tree is a user-defined data structure that helps to calculate Prefix sums or cumulative frequency tables in a very efficient manner.It is a representational form of an array and can query or update elements in O(logN) time complexity.It is also known as a Binary Indexed Tree as it makes use of bit manipulation concepts for implementation purposes.
```
Comment (Suvraneel) [inline C_CPP/Data Structures/User Defined Data Structures/Fenwick Tree/Fenwick_Tree.md:3]:
```suggestion
A Fenwick Tree is a user-defined data structure that helps to calculate Prefix sums or cumulative frequency tables in a very efficient manner. It is a representational form of an array and can query or update elements in **O(logN)** time complexity. It is also known as a Binary Indexed Tree as it makes use of bit manipulation concepts for implementation purposes.
```

File: C_CPP/Data Structures/User Defined Data Structures/Fenwick Tree/Fenwick_Tree.md [hunk 5]
```diff
@@ -0,0 +1,52 @@
...
+- We can perform both Query and Update operations even in a large array in O(logN) time using Fenwick Trees.
+- Fenwick Trees require less space as compared to Segment Tress.
+- They have an easy implementation code which is useful during various programming contests.
+
+Terminology:
+
+- Query:Operation performed on an interval or on an element.For example minimum or maximum or summation in a particular range.
+- Update:To update the fenwick tree.There are two types of updates in a fenwick tree.
+  - Point Updates:Updating a particular element of the binary indexed tree with a value.
+  - Range Updates:Executing queries on a range of elements with a value.
```
Comment (Suvraneel) [inline C_CPP/Data Structures/User Defined Data Structures/Fenwick Tree/Fenwick_Tree.md:16]:
```suggestion
- **Query**: Operation performed on an interval or on an element.For example minimum or maximum or summation in a particular range.
- **Update**: To update the Fenwick tree. There are two types of updates in a Fenwick tree.
  - **Point Updates**: Updating a particular element of the binary indexed tree with a value.
  - **Range Updates**: Executing queries on a range of elements with a value.
```

Not requesting same change below this... Kindly format/adorn the doc so it is more legible than plain text

File: C_CPP/Data Structures/User Defined Data Structures/Fenwick Tree/Fenwick_Tree.md [hunk 6]
```diff
@@ -0,0 +1,52 @@
+<h1 align="center">  Fenwick Tree or Binary Indexed Tree </h1>
+
+A Fenwick Tree is a user-defined data structure that helps to calculate Prefix sums or cumulative frequency tables in a very efficient manner.It is a representational form of an array and can query or update elements in O(logN) time complexity.It is also known as a Binary Indexed Tree as it makes use of bit manipulation concepts for implementation purposes.
+
+Need:
+
+- We can perform both Query and Update operations even in a large array in O(logN) time using Fenwick Trees.
+- Fenwick Trees require less space as compared to Segment Tress.
```
Comment (Suvraneel) [inline C_CPP/Data Structures/User Defined Data Structures/Fenwick Tree/Fenwick_Tree.md:8]:
```suggestion
- Fenwick Trees require less space as compared to Segment Trees.
```

File: C_CPP/Data Structures/User Defined Data Structures/Fenwick Tree/Fenwick_Tree.md [hunk 7]
```diff
@@ -0,0 +1,52 @@
...
+Terminology:
+
+- Query:Operation performed on an interval or on an element.For example minimum or maximum or summation in a particular range.
+- Update:To update the fenwick tree.There are two types of updates in a fenwick tree.
+  - Point Updates:Updating a particular element of the binary indexed tree with a value.
+  - Range Updates:Executing queries on a range of elements with a value.
+
+## Bit Manipulation Concepts:
+
+The basic concept behind Fenwick Tree is that we can represent all positive integers as the sum of powers of 2.Example: 21 can be represented as 21 = 16 + 4 + 1 = 2^4 +2^2 +2^0.This is the reason why Fenwick Tree is also referred to as the Binary Indexed Tree.
```
Comment (Suvraneel) [inline C_CPP/Data Structures/User Defined Data Structures/Fenwick Tree/Fenwick_Tree.md:20]:
```suggestion
The basic concept behind Fenwick Tree is that we can represent all positive integers as the sum of powers of 2.  
Example: 21 can be represented as 21 = 16 + 4 + 1 = 2<sup>4</sup> +2<sup>2</sup> +2<sup>0</sup>.  
This is the reason why Fenwick Tree is also referred to as the Binary Indexed Tree.
```

Comment (Suvraneel) [reply C_CPP/Data Structures/User Defined Data Structures/Fenwick Tree/Fenwick_Tree.md:20]: @shreya024  This one is optional though...
If you don't feel like using sup tags then just embolden the numerical text

File: C_CPP/Data Structures/User Defined Data Structures/Fenwick Tree/Fenwick_Tree.md [hunk 8]
```diff
@@ -0,0 +1,52 @@
...
+
+  - L.S.B is the Least Significant bit of the binary representation of any given number.
+    ![image](https://user-images.githubusercontent.com/72400676/139364185-263bbadf-237d-4d01-8f7b-1f3c1e12bae4.png)
+  - Here we observe a chain of responsibilities i.e. when we drop the L.S.B from 11,we get 10.
+  - Progressively dropping the chain of L.S.B s, we can reach 0.
+  - Similarly Adding up the L.S.B s will help us to propagate the value up to the cells responsible.
+
+- Isolation of L.S.B
+  - Let a positive number be of the form A1B where A contains some random binary bits and B contains only zeroes.
+  - Taking 2's complement of A1B: (A1B)'+1= (A'0B')+1
```
Comment (Suvraneel) [inline C_CPP/Data Structures/User Defined Data Structures/Fenwick Tree/Fenwick_Tree.md:32]: Bolden

File: C_CPP/Data Structures/User Defined Data Structures/Fenwick Tree/Fenwick_Tree.md [hunk 9]
```diff
@@ -0,0 +1,52 @@
...
+  - Taking bitwise AND : A1B & A'1B' = A number of the form 0....1....0
+    For example:
+    ![image](https://user-images.githubusercontent.com/72400676/139364135-1bfc2a26-e9d9-415a-b30c-fe5bec458b3c.png)
+    In this way, we can extract the Least Significant bit from the binary representation of any number to carry our different operations.
+
+## Algorithm:
+
+- Query:In each iteration we are removing a L.S.B from the given number till we reach 0.
+- Point Updates: In each iteration we are adding a L.S.B to the number.
+- Range Updates: We are adding val to all elements from L to end and -val to all elements from R till end such that val is added only to elements from L to R.
```
Comment (Suvraneel) [inline C_CPP/Data Structures/User Defined Data Structures/Fenwick Tree/Fenwick_Tree.md:43]: Bolden headings

File: C_CPP/Data Structures/User Defined Data Structures/Fenwick Tree/Fenwick_Tree.md [hunk 10]
```diff
@@ -0,0 +1,52 @@
...
+
+The number of set bits in the binary representation of a number n is O(logN).We at worst case,the number of traversals is O(logN).So,the time complexity is O(logN).
```
Comment (Suvraneel) [inline C_CPP/Data Structures/User Defined Data Structures/Fenwick Tree/Fenwick_Tree.md:45]:
```suggestion
The number of set bits in the binary representation of a number n is O(logN). We at worst case, the number of traversals is **O(logN)**. So, the time complexity is**O(logN)**.
```

File: C_CPP/Data Structures/User Defined Data Structures/Fenwick Tree/Fenwick_Tree.md [hunk 11]
```diff
@@ -0,0 +1,52 @@
...
+
+## Sample Problem:
+
+Question: Take an input array from the user.
+Operations:1.Print the Sum of Array from Range L to R.
+2.Update the value of an element of given index with val.
+3.Add val to each element of the array from range L to R.
```
Comment (Suvraneel) [inline C_CPP/Data Structures/User Defined Data Structures/Fenwick Tree/Fenwick_Tree.md:52]:
```suggestion
Question: Take an input array from the user.  
Operations:  
1.Print the Sum of Array from Range L to R.  
2.Update the value of an element of given index with val.  
3.Add val to each element of the array from range L to R.  
```

Add 2 spaces at end of prev line to treat as new line

**Review by Suvraneel — CHANGES_REQUESTED** [review]
@shreya024  Great topic selection & nice content btw 😄

File: C_CPP/Data Structures/User Defined Data Structures/Fenwick Tree/Fenwick_Tree.md [hunk 12]
```diff
@@ -0,0 +1,52 @@
...
+  - Similarly Adding up the L.S.B s will help us to propagate the value up to the cells responsible.
+
+- Isolation of L.S.B
+  - Let a positive number be of the form A1B where A contains some random binary bits and B contains only zeroes.
+  - Taking 2's complement of A1B: (A1B)'+1= (A'0B')+1
+  - Now the number would be of the form A'1B'
+  - Taking bitwise AND : A1B & A'1B' = A number of the form 0....1....0
+    For example:
+    ![image](https://user-images.githubusercontent.com/72400676/139364135-1bfc2a26-e9d9-415a-b30c-fe5bec458b3c.png)
+    In this way, we can extract the Least Significant bit from the binary representation of any number to carry our different operations.
```
Comment (Suvraneel) [inline C_CPP/Data Structures/User Defined Data Structures/Fenwick Tree/Fenwick_Tree.md:37]:
```suggestion
    In this way, we can extract the Least Significant bit from the binary representation of any number to carry out different operations.
```
Typo... Run through Grammarly if needed

**Comment by shreya024** [comment]
@Suvraneel Thank you and I have updated the changes suggested. Kindly check and let me know.

**Comment by Suvraneel** [comment]
@shreya024  LGTM ❤️   Awesome  GG 🚀

**Pull request merged**